packages/cli/src/commands/i18n/i18nScan.ts

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import fs from 'fs';
+import glob from 'glob';
+import i18n from 'i18next-scanner';
+import { OptionValues } from 'commander';
+
+export default async (opts: OptionValues) => {
+  const parser = new i18n.Parser({
```

Comment:
Tbh I think we can do something much better than what this scanner is able to do. We have the `TranslationRef`s with default messages and potentially event comments for those. It provides both much more context for the translations, but also keeps them grouped.

I'm thinking it'd be able to use something like `ts-morph` to find all references of `createTranslationRef` in the project and extract the messages?

Reply:
I can try it out!

Reply:
I played with the library and here is what I found. I used [this file](https://github.com/backstage/backstage/blob/30e76f392e3cb84071438c13088d5bec8d4ea92f/plugins/user-settings/src/translation.ts#L20) and tried to get to the `messages` prop.

So here is the code:
```ts
const project = new Project({
    tsConfigFilePath: `./tsconfig.json`,
  });
  const sourceFile = project.addSourceFileAtPath(
    './plugins/user-settings/src/translation.ts',
  );
  sourceFile.getVariableDeclarations().forEach(variableDeclaration =>
    variableDeclaration
      .getChildrenOfKind(SyntaxKind.CallExpression)
      .forEach(child =>
        child.getArguments().forEach(arg =>
          arg
            .asKind(SyntaxKind.ObjectLiteralExpression)
            ?.getProperties()
            .forEach((prop, idx) =>
              console.log(
                `\nProp #${idx}: ${prop
                  .asKind(SyntaxKind.PropertyAssignment)
                  ?.getText()}`,
              ),
            ),
        ),
      ),
  );
```

This will end up printing something like:

```ts
Prop #0: id: 'user-settings'

Prop #1: messages: {
    languageToggle: {
      title: 'Language',
      description: 'Change the language',
      select: 'Select language {{language}}',
    },
    themeToggle: {
      title: 'Theme',
      description: 'Change the theme mode',
      select: 'Select theme {{theme}}',
      selectAuto: 'Select Auto Theme',
      names: {
        light: 'Light',
        dark: 'Dark',
        auto: 'Auto',
      },
    },
  }
```

From this I could recursively map through the object and eventually get all the keys/values pairs. The challenge will be if there is a spread operator used inside of the object, as I won't be able to figure this one out.

Let me know if this is the approach I should go with and I can update the PR! (cc @awanlin )

Reply:
Another approach could be that each `translation.ts` file exports the object and we could easily import it. The implementation would be much simpler but it would require changes to the codebase.

Reply:
Perhaps the cleanes implementation would be to get hold of the `TranslationRef` type and enumerate a all exported instances of it, if that's possible. If we can avoid requiring a particular file structure in plugins that's definitely desirable. I think what you got so far looks good though, it would indeed not be too tricky to restructure that into the actual message keys. Identifying all occurrences would be the hard part at this point.

It's kinda tricky to wire that all up though so feeling punting on this work a bit would be reasonable too.